### PR TITLE
create and use a more richly typed snapd api

### PR DESCRIPTION
**Pull request by mwhudson**
Some bits of this should probably be split off and maybe a little rebasing should be done but I'm interested in what you think about this.

### CODE REVIEW

**Comment by dbungert** [comment]
I'll want to read in more detail but at first glance seems straightforward.
The type annotation stuff is a bit mysterious.

File: subiquity/server/controllers/refresh.py [hunk 1]
```diff
@@ -213,16 +221,14 @@ async def check_for_update(self, context):
...
-    async def get_progress(self, change):
-        result = await self.app.snapd.get('v2/changes/{}'.format(change))
-        change = result['result']
-        if change['status'] == 'Done':
+        change_id = await self.app.snapdapi.v2.snaps[self.snap_name].POST(
+            SnapActionRequest(action=SnapAction.REFRESH, ignore_running=True))
+        context.description = "change id: {}".format(change_id)
+        return change_id
+
+    async def get_progress(self, change_id):
```
Comment (dbungert) [inline subiquity/server/controllers/refresh.py:229]: Suggest annotating this one.

File: subiquity/server/server.py [hunk 2]
```diff
@@ -313,9 +314,11 @@ def __init__(self, opts, block_log_dir):
                     "examples", "snaps"),
                 self.scale_factor, opts.output_base)
             self.snapd = AsyncSnapd(connection)
+            self.snapdapi = make_api_client(self.snapd)
         elif os.path.exists(self.snapd_socket_path):
             connection = SnapdConnection(self.root, self.snapd_socket_path)
             self.snapd = AsyncSnapd(connection)
+            self.snapdapi = make_api_client(self.snapd)
         else:
             log.info("no snapd socket found. Snap support is disabled")
```
Comment (dbungert) [inline subiquity/server/server.py:323]: For the current support this is appropriate but that might need to be upgraded to a more serious error later on.

**Review by dbungert — COMMENTED** [review]
The changes are pretty straightforward.  Just my previous comment on some of the type assertion changes and a few more comments.

File: subiquity/server/controllers/refresh.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/refresh.py:229]: done

File: subiquity/server/server.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/server/server.py:323]: this is here because the wsl system setup code comes through here (we need to fix this somehow or other!)

**Comment by mwhudson** [comment]
> The type annotation stuff is a bit mysterious.

Do you mean the stuff I have now split out into a separate PR? (#1440) Or something else?

**Comment by dbungert** [comment]
> Do you mean the stuff I have now split out into a separate PR?

Yes

**Review by dbungert — APPROVED** [review]

**Comment by mwhudson** [comment]
ahh poo this breaks in a vm :( i think the snapd output has changed a bit.

**Pull request merged**